enterprise/internal/batches/sources/perforce.go

```diff
@@ -0,0 +1,178 @@
+		}
+	default:
+		return s, errors.New("unexpected auther type for Perforce Source")
+	}
+
+	return s, nil
+}
+
+// ValidateAuthenticator validates the currently set authenticator is usable.
+// Returns an error, when validating the Authenticator yielded an error.
+func (s PerforceSource) ValidateAuthenticator(ctx context.Context) error {
+	if s.perforceCreds == nil {
+		return errors.New("no credentials set for Perforce Source")
+	}
+	rc, _, err := s.gitServerClient.P4Exec(ctx, s.perforceCreds.Host, s.perforceCreds.Username, s.perforceCreds.Password, "users")
+	if err == nil {
+		_ = rc.Close()
+		return nil
+	}
+	return err
+}
+
+// LoadChangeset loads the given Changeset from the source and updates it. If
+// the Changeset could not be found on the source, a ChangesetNotFoundError is
+// returned.
+func (s PerforceSource) LoadChangeset(ctx context.Context, cs *Changeset) error {
+	if s.perforceCreds == nil {
+		return errors.New("no credentials set for Perforce Source")
+	}
+	cl, err := s.gitServerClient.P4GetChangelist(ctx, cs.ExternalID, *s.perforceCreds)
+	if err != nil {
+		return errors.Wrap(err, "getting changelist")
+	}
+	return errors.Wrap(s.setChangesetMetadata(cl, cs), "setting perforce changeset metadata")
+}
+
+// CreateChangeset will create the Changeset on the source. If it already
+// exists, *Changeset will be populated and the return value will be true.
+func (s PerforceSource) CreateChangeset(ctx context.Context, cs *Changeset) (bool, error) {
+	return false, s.LoadChangeset(ctx, cs)
+}
+
+// CreateDraftChangeset creates the given changeset on the code host in draft mode.
+// Perforce does not support draft changelists
+func (s PerforceSource) CreateDraftChangeset(_ context.Context, _ *Changeset) (bool, error) {
+	return false, errors.New("not implemented")
+}
+
+func (s PerforceSource) setChangesetMetadata(cl *protocol.PerforceChangelist, cs *Changeset) error {
+	if err := cs.SetMetadata(cl); err != nil {
+		return errors.Wrap(err, "setting changeset metadata")
+	}
+
+	return nil
+}
+
+// UndraftChangeset will update the Changeset on the source to be not in draft mode anymore.
+func (s PerforceSource) UndraftChangeset(_ context.Context, _ *Changeset) error {
+	// TODO: @peterguy implement this function?
+	// not sure what it means in Perforce - submit the changelist?
+	return errors.New("not implemented")
+}
+
+// CloseChangeset will close the Changeset on the source, where "close"
+// means the appropriate final state on the codehost.
+// deleted on Perforce, maybe?
+func (s PerforceSource) CloseChangeset(_ context.Context, _ *Changeset) error {
+	// TODO: @peterguy implement this function
+	// delete changelist?
+	return errors.New("not implemented")
+}
+
+// UpdateChangeset can update Changesets.
+func (s PerforceSource) UpdateChangeset(_ context.Context, _ *Changeset) error {
+	// TODO: @peterguy implement this function
+	// not sure what this means for Perforce
+	return errors.New("not implemented")
+}
```

Comment:
This is probably a no-op since any updates (changes to the title, "commit" message, author, etc.) would all happen from gitserver with the p4 commands, right?

Probably here you would just want to make sure to reload the changeset so we've got the most recent metadata recorded for it:

```suggestion
// UpdateChangeset can update Changesets.
// Noop since Perforce pushes updates directly with p4
func (s PerforceSource) UpdateChangeset(ctx context.Context, cs *Changeset) error {
	return s.LoadChangeset(ctx, cs)
}
```